File: dotnet/Platforms/OpenStack.EginX/Platform_EginX.cs
using EginX;
using OpenStack.Client;
using OpenStack.Gfx;
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
#pragma warning disable CS0649, CS0169

[assembly: InternalsVisibleTo("OpenStack.GfxTests")]

namespace OpenStack;

#region Client

public class ExClientHost : Game, IClientHost {
    public ClientBase Client;
    public SceneBase Scene;
    public IPluginHost PluginHost;

    public ExClientHost(Func<ClientBase> client) {
        Client = client();
        DeviceManager = new GraphicsDeviceManager(this);
        //PluginHost = pluginHost; IPluginHost pluginHost, string title
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public T GetScene<T>() where T : SceneBase => Scene as T;

    public void SetScene(SceneBase scene) { Scene?.Dispose(); Scene = scene; Scene?.Load(); }

    protected override async Task LoadContent() {
        await base.LoadContent();
        await Client.LoadContent();
    }

    protected override async Task UnloadContent() {
        await Client.UnloadContent();
        await base.UnloadContent();
    }
}


#endregion

#region Platform

// EginXObjectBuilder
// MISSING

// EginXGfxSprite2D
public class EginXGfxSprite2D : IOpenGfxSprite<object, object> {
    readonly ISource _source;
    readonly SpriteManager<object> _spriteManager;
    public EginXGfxSprite2D(ISource source) {
        _source = source;
        //_spriteManager = new SpriteManager<Sprite2D>(source, new GodotSpriteBuilder());
    }

    public ISource Source => _source;
    public SpriteManager<object> SpriteManager => _spriteManager;
    public void PreloadSprite(object path) => _spriteManager.PreloadSprite(path);
    public Task<(object spr, object tag)> CreateSprite(object path, object parent = null) => _spriteManager.CreateSprite(path);
}

// EginXPlatform
public class EginXPlatform : Platform {
    public static Dictionary<Type, Func<object, bool, object, object>> BuildersByType = [];
    public static readonly Platform This = new EginXPlatform();
    EginXPlatform() : base("EX", "EginX") {
        GfxFactory = source => [null, new EginXGfxSprite2D(source), null, null, null, null];
        SfxFactory = source => [new SystemSfx(source)];
    }
}

#endregion

File: Python/openstk/platforms/eginx/platform_eginx.py
from __future__ import annotations
import traceback
from openstk.core import ISource, Platform
from openstk.platforms.system import SystemSfx
from openstk.gfx import IOpenGfxApi, IOpenGfxSprite, IOpenGfxModel, ObjectModelBuilderBase, ObjectModelManager, MaterialBuilderBase, MaterialManager, ShaderBuilderBase, ShaderManager, TextureManager, TextureBuilderBase
from openstk.client import IClientHost
from openstk.gfx.egin import Game, GraphicsDeviceManager


#region Client

# EginXClientHost
class EginXClientHost(Game, IClientHost):
    def __init__(self, client: callable):
        super().__init__()
        # self.deviceManager = GraphicsDeviceManager(self)
        self.client: ClientBase = client()
        self.scene: SceneBase = None
        self.pluginHost: IPluginHost = None

    def getScene[T](self) -> T: return self.scene

    def setScene(self, scene: SceneBase) -> None:
        if self.scene: self.scene.dispose()
        self.scene = scene
        self.scene.load() 

    def loadContent(self) -> None:
        super().loadContent()
        self.client.loadContent()

    def unloadContent(self) -> None:
        self.client.unloadContent()
        super().unloadContent()

#endregion

#region Platform

# EginXObjectBuilder
# MISSING

# EginXGfxApi
class EginXGfxApi(IOpenGfxApi):
    def __init__(self): pass
    def attach(self, method: GfxAttach, src: object, args: list[object]) -> object: raise NotImplementedError()

# EginXGfxSprite2D
class EginXGfxSprite2D(IOpenGfxSprite):
    def __init__(self):
        # self.spriteManager: SpriteManager = SpriteManager(OpenGLTextureBuilder())
        # self.objectManager: ObjectSpriteManager = ObjectManager(OpenGLObjectModelBuilder())
        pass
    # def preloadObject(self, source: ISource, path: object) -> None: raise NotImplementedError()
    def preloadSprite(self, source: ISource, path: object) -> None: self.textureManager.spriteManager(path)
    # def createObject(self, source: ISource, path: object, parent: object) -> tuple[object, dict[str, object]]: raise NotImplementedError()
    def createSprite(self, source: ISource, path: object, level: range = None) -> int: return self.spriteManager.createSprite(path)[0]

# EginXPlatform
class EginXPlatform(Platform):
    def __init__(self):
        super().__init__('EX', 'EginX')
        self.gfxFactory = staticmethod(lambda: [EginXGfxApi(), EginXGfxSprite2D(), None, None, None, None])
        self.sfxFactory = staticmethod(lambda: [SystemSfx()])
EginXPlatform.this = EginXPlatform()

#endregion